### ISSUE TITLE
Gradient disappear when there is an indexed-array in the loss function

### ISSUE THREAD

**Issue by huangyxi** [FluxML/Flux.jl#1970]
Hi, I am going to train some parameters specified by labels in a neural network. I found that when using an indexed array in the loss function, the corresponding gradient appears to be `nothing`. Here is a minimal working example:
```julia
using Flux

x1 = [0., 1.]
x = [x1]
x_real = x[1]
p = [0.5, 0.6]

function loss()
    y = x[1] + p # line {1}
    sum(abs2, y .- 1)
end

ps = Flux.params(x[1], p)
gs = Flux.gradient(loss, ps)
```
I have tried that the gradient descent is performed normally whether line `{1}` be `y = x[1] # {2}` or `y = x_real + p # {3}`. However, when line `{1}` is `y = x[1] + p` (`# {2}` or `# {3}` is not suitable for our condition), the results of the gradient `gs.grads` is shown below:
```julia
IdDict{Any, Any} with 4 entries:
  [0.0, 1.0] => nothing
  [0.5, 0.6] => [-1.0, 1.2]
  :(Main.x)  => [[-1.0, 1.2]]
  :(Main.p)  => [-1.0, 1.2]
```
which indicates that the corresponding gradient of `x1` is not been performed.
